### ISSUE TITLE
Feature request: Docker API via TCP

### ISSUE THREAD

**Issue by sammcj** [capnspacehook/whalewall#133]
Howdy, nice project!

Have you got any thoughts on defaulting to use something such as https://github.com/Tecnativa/docker-socket-proxy to further secure access to the docker daemon?

I use it with Traefik and have found it really reliable.

I think all that would be required is allowing access to the Docker via TCP rather than a unix socket, then you could do something like:

```yaml
networks:
  docker-proxynet: # Used for securely exposing docker.sock
    driver: bridge
    name: docker-proxynet
    internal: true
    enable_ipv6: false

services:
  dockerproxy: # see https://github.com/Tecnativa/docker-socket-proxy
    environment:
      - CONTAINERS=1
      - NETWORKS=1 (maybe?)
    container_name: dockerproxy
    image: ghcr.io/tecnativa/docker-socket-proxy:edge
    cpus: 0.75
    hostname: dockerproxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    networks:
      - docker-proxynet # use only internal network
    ports:
      - 127.0.0.1:2375:2375

  whalewall:
    cap_add: 
      - NET_ADMIN
    image: ghcr.io/capnspacehook/whalewall
    network_mode: host
    volumes:
      - whalewall_data:/data
    environment:
      - DOCKER_API: tcp://dockerproxy:2375

volumes:
  whalewall_data:
```

**Comment by capnspacehook** [comment]
Hey, I've actually looked into that project but eventually came to the conclusion that using it wouldn't really improve security. I say this because in this case, docker-socket-proxy and whalewall have almost the same permissions. Both (AFAIK) don't connect to anything other than a Docker daemon. Whalewall does additionally have the CAP_NET_ADMIN capability, but I don't think that greatly adds to the risk.

Basically, using docker-socket-proxy is just choosing which image gets unrestricted Docker daemon access (since it can't securely limit Docker access to itself). I don't think that adds much security wise, not enough for me to use it. If whalewall started making connections to the internet as a part of its normal behavior, then I would consider its attack surface expanded to the point where using docker-socket-proxy would make more sense.

Feel free to use docker-socket-proxy yourself though, that's just my thoughts on it. You should be able to right now by setting some environmental variables:

```
DOCKER_HOST to set the URL to the docker server.

DOCKER_API_VERSION to set the version of the API to use, leave empty for latest.

DOCKER_CERT_PATH to specify the directory from which to load the TLS certificates (ca.pem, cert.pem, key.pem).

DOCKER_TLS_VERIFY to enable or disable TLS verification (off by default).
```

Specifically DOXKER_HOST is probably what you need. I'll add these to the README so they're documented.

**Comment by sammcj** [comment]
Ah, I'm glad you've looked into it! Thank you for the detailed response I really appreciate (and the envs!).

**Comment by capnspacehook** [comment]
No problem! Going to reopen this to remind me to add this info to the README

**Comment by capnspacehook** [comment]
Closed via df64903382e41789e74a8b72c90656f4ef7a4ac0.
